### DISCUSSION TITLE
Affect the Visibility for Swift Structs Generated from Protobuf

### DISCUSSION THREAD

**Discussion by cgrindel** [rules-proto-grpc/rules_proto_grpc#97] [Q&A] [answered]
In other words, I would like my generated structs to be `public`.

I noticed that the swift-protobuf plugin for protoc supports a [Visibility](https://github.com/apple/swift-protobuf/blob/main/Documentation/PLUGIN.md#generation-option-visibility---visibility-of-generated-types) option. However, I am not sure how to configure the `swift_proto_compile` rule to specify this. I assumed that there would be an attribute on the rule or an `option` that could be specified in the proto file. Unfortunately, my investigation has not turned up an answer. 

Any help would be greatly appreciated. Thanks.

**Comment by aaliddell** [comment, marked as answer]
There isn't presently a simple way of adding options on a per-target basis, due to the way aspects work in Bazel. You can add the options on a per-plugin basis, but this would require you to make your own copy of the plugin, which is a pain. The relevant issue is #54 

I'm going to add another issue so we can track this for swift. Since this is only one option, it may be viable to work around the aspect limitation since we can enumerate the values.

Another option would be just to make the public visibility the default option for Swift, but that may have other implications. This is what `rules_swift` does: https://github.com/bazelbuild/rules_swift/blob/ca8fe5b39c273d1368723ec7e88470365c173687/swift/internal/swift_grpc_library.bzl#L122

**Reply by aaliddell** [reply]
Tracking issue is #111

**Reply by cgrindel** [reply]
Thanks for the response. I ended up creating a patch file that adjusts the visibility to public.

**Reply by aaliddell** [reply]
Ok, is there a disadvantage to just making public the default?

I could add this to 3.0.0, but don't know if I'll end up breaking people's builds if I do. As far as I can see it won't, but my experience with Swift is limited.

**Reply by cgrindel** [reply]
I don't think so. Given that protobuf structures tend to be used for serialization and interoperability, I think setting the visibility to public makes sense.

**Reply by aaliddell** [reply]
Ok, thanks for the details, I'll do that shortly :+1:
